internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,52 +100,210 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
+	switch *upstream.Spec.Claims.Username {
+	case v1alpha1.GitHubUsernameLoginAndID:
+	case v1alpha1.GitHubUsernameLogin:
+	case v1alpha1.GitHubUsernameID:
+		break
+	default:
+		return nil, fmt.Errorf("invalid spec.claims.username")
+	}
+
+	// Should not happen due to CRD enum validation
+	switch *upstream.Spec.Claims.Groups {
+	case v1alpha1.GitHubUseTeamNameForGroupName:
+	case v1alpha1.GitHubUseTeamSlugForGroupName:
+		break
+	default:
+		return nil, fmt.Errorf("invalid spec.claims.groups")
+	}
+
+	var policy v1alpha1.GitHubAllowedAuthOrganizationsPolicy
+	if upstream.Spec.AllowAuthentication.Organizations.Policy != nil {
+		policy = *upstream.Spec.AllowAuthentication.Organizations.Policy
+	}
+
+	// Should not happen due to CRD defaulting, enum validation, and CEL validation (for recent versions of K8s only!)
+	invalidPolicyCondition := &metav1.Condition{
+		Type:    OrganizationsPolicyValid,
+		Status:  metav1.ConditionFalse,
+		Reason:  "Invalid",
+		Message: `allowAuthentication policy for organizations is not valid`,
+	}
+	switch policy {
+	case v1alpha1.GitHubAllowedAuthOrganizationsPolicyAllGitHubUsers:
+		if len(upstream.Spec.AllowAuthentication.Organizations.Allowed) > 0 {
+			conditions = append(conditions, invalidPolicyCondition)
+		}
+		break
+	case v1alpha1.GitHubAllowedAuthOrganizationsPolicyOnlyUsersFromAllowedOrganizations:
+		if len(upstream.Spec.AllowAuthentication.Organizations.Allowed) == 0 {
+			conditions = append(conditions, invalidPolicyCondition)
+		}
+		break
+	default:
+		conditions = append(conditions, invalidPolicyCondition)
+	}
+
+	result := &upstreamgithub.ProviderConfig{
+		Name:               upstream.Name,
+		ResourceUID:        upstream.UID,
+		Host:               fmt.Sprintf("https://" + host),
+		GroupNameAttribute: *upstream.Spec.Claims.Groups,
+		UsernameAttribute:  *upstream.Spec.Claims.Username,
+		Config: &oauth2.Config{
+			ClientID:     clientID,
+			ClientSecret: clientSecret,
+		},
+		AllowedOrganizations:    upstream.Spec.AllowAuthentication.Organizations.Allowed,
+		OrganizationLoginPolicy: policy,
+		AuthorizationURL:        fmt.Sprintf("https://%s/login/oauth/authorize", host),
 	}
 
 	// TODO: once we firm up the proposal doc & merge, then firm up the CRD & merge, we can fill out these validations.
 	// The critical pattern to maintain is that every run of the sync loop will populate the exact number of the exact
 	// same set of conditions.  Conditions depending on other conditions should get Status:  metav1.ConditionUnknown, or
 	// Status:  metav1.ConditionFalse, never be omitted.
-	conditions := []*metav1.Condition{
-		// we may opt to split this up into smaller validation functions.
-		// Each function should be responsible for validating one logical unit and setting one condition.
-		// c.validateGitHubAPI(),
-		// c.validateClaims(),
-		// c.validateAllowedAuthentication(),
-		// c.validateClient(),
+	if !hasConditionByType(conditions, TLSConfigurationValid) {
```

Comment:
For long term maintenance, I tend to think keeping all the branches for one condition together is ideal (ie, a helper func).  Its a little difficult to jump around and keep track of each `Type` mentally, and match them up (is it one branch, two, or three that sets `Type: foo`?).

Reply:
one helper function per condition now